### PR TITLE
Getting papers from Crossref API

### PR DESCRIPTION
**Pull request by davin111**
Related Issue: #73, #82

# Major changes
## Backend
When users search, now papers are retrieved also from [Crossref](https://github.com/CrossRef/rest-api-doc#crossref-rest-api), not only from arXiv.

For optimizing search speed, I used two threads. One calls arXiv API and the other calls Crossref API.

Backend gives at most 40 papers to frontend, if the two external APIs give results properly. For now, frontend show all papers from backend at once. Surely, users can request more papers by clicking the 'View More' button. If users click it, the next pages is requested for arXiv and Crossref.

If something's wrong in communicating with external APIs, our search API just give results by naive-searching(string match) in our DB. In this case, backend gives at most 20 papers.

## Frontend
스크린샷 2019-11-30 17 45 28

SearchResult spreads PaperCards of given papers, considering their source.

# Minor changes
I replaced `print` with `logging` functions in Django codes. We can easily switch between logging modes by modifying one line in `settings.py`.

I fixed bugs that users can't search by pressing 'enter' key on SearchResult page.

# Remaining problems
In theory, papers from Crossref don't have problem. However, most papers from Crossref don't have abstracts, which are quite important parts of meta information. Also, we can't extract keywords when there is no abstract.

I found another external API which can help it. [SemanticScholar API](https://api.semanticscholar.org) gives quite rich information(including abstract) of one paper. However, there is another problem. It gives the information only when we give the specific ID(ex> DOI) of papers to it! 

It means we should have at least two steps for getting those information. One step is getting basic information(including DOI) from Crossref, and the other step is getting full information(including abstract) from SemanticScholar, giving the DOI.

This is the process which this PR use.
```
Users search -> (thread 1) arXiv API       -> (join) give response to users
                -> (thread 2) Crossref API ->
```

I'm thinking about this process.

Celery(http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html#django-first-steps) is a library helping **asynchronous jobs**(meaning, users don't need to wait the jobs). Also, it can be used with Django.
```
Users search -> (thread 1) arXiv API       -> (join) give response to users
                -> (thread 2) Crossref API ->
                                         -> trigger shared_task in [Celery] (from here, this process is independent of search response)
                                                -> SemanticScholar APIs for each results of Crossref, not including abstracts)
                                                     -> update DB
```

**However, there are two potential problems with this.**
First, users can't see keywords of PaperCards on SearchResult because keywords are extracted from abstracts.
Second, if users search some papers and click the PaperCard quickly, users will be redirected to the PaperDetail page. However in that time, detail information of the paper may be not in our DB yet.

**Please ask any questions here or our Slack workspace, because this problem is important and quite complex!**

### Checklist
This pull request...
- [x] has preceding issues related with it (if resolves some issues, metions them with 'resolve')
- [x] doesn't include too many changes (you should focus on the specific feature in a PR)
- [x] elaborately explains what feature is added, which part is fixed or improved
- [x] has step by step instructions for collaborators, if needed
- [x] has attached screenshots, if appropriate

My codes...
- [x] include enough tests for added and changed parts
- [x] pass all the new and existing tests
- [x] include appropriate comments
- [x] don't introduce unnecessary warnings

### CODE REVIEW

File: backend/PapersFeed_backend/settings.py [hunk 1]
```diff
@@ -134,3 +137,11 @@
 
 # CSRF
 CSRF_COOKIE_NAME = 'csrftoken'
+
+# CELERY
```
Comment (davin111) [inline backend/PapersFeed_backend/settings.py]: These options will be used in the next PR.

File: backend/papersfeed/utils/papers/utils.py [hunk 2]
```diff
@@ -99,41 +103,56 @@ def select_paper_search(args):
...
+        'arxiv': [],
+        'crossref': [],
+    }
+    is_finished_dict = {
+        'arxiv': False,
+        'crossref': False,
+    }
+
+    future_to_source = {}
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```
Comment (davin111) [inline backend/papersfeed/utils/papers/utils.py]: You can refer to [here](https://docs.python.org/ko/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) for these codes.

**Comment by davin111** [comment]
@kosumin You mean, how long it takes until detailed information of searched papers are saved in our DB through SemanticScholar API?

By the way, I checked multithreading is good for performance. The time duration of dividing flow to arXiv and Crossref and joining them almost equals to a single call for arXiv API. Also, I added timeout(2 seconds) in the codes.

**Comment by kosumin** [comment]
No I mean how long it would take to save information from SemanticScholar to DB. I think hopefully there could be delays between front and back or front works and the info could be saved during the delays (before user see Paper Detail Page).

**Comment by davin111** [comment]
Yeah, I asked about it! Well, hopefully, I want it works. Using Celery is quite challenging, and I don't want situations throwing out related implementation. :( Anyway, I'll try it after @abcthing and @njsh4261 check this PR.

**Comment by davin111** [comment]
Also, if there is a good external source(API) which supports search query and gives enough information of papers, surely we'll use it! Please let me know if you find something. We might get a clue from https://libraries.mit.edu/scholarly/publishing/apis-for-scholarly-resources/.

**Comment by abcthing** [comment]
Approve

**Pull request merged**